reflect/src/main/java/dagger/reflect/Scope.java

```diff
@@ -65,30 +86,52 @@ private Scope(BindingMap bindings, JustInTimeLookup.Factory jitLookupFactory,
    * Otherwise, the parent scope (if any) will be recursively checked. If no matching scope
    * annotation is found by traversing the parents null will be returned.
    */
-  private @Nullable LinkedBinding<?> putJitBinding(Key key, JustInTimeLookup lookup) {
+  private @Nullable LinkedBinding<?> putJitBinding(Key key, @Nullable Linker linker,
+      JustInTimeLookup lookup) {
     Binding jitBinding = lookup.binding;
 
     Annotation scope = lookup.scope;
     if (scope != null) {
       if (!annotations.contains(scope)) {
         return parent != null
-            ? parent.putJitBinding(key, lookup)
+            ? parent.putJitBinding(key, linker, lookup)
             : null;
       }
 
       jitBinding = jitBinding.asScoped();
     }
 
-    Binding binding = bindings.putIfAbsent(key, jitBinding);
+    Binding replaced = bindings.putIfAbsent(key, jitBinding);
+    Binding binding = replaced != null
+        ? replaced // You raced another thread and lost.
```

Comment:
just curious, why don't we care about thread races here? because they aren't singletons?

Reply:
These races are only to create the binding, not the instance.

Two threads might both perform the reflection to find the `@Inject` constructor on some type at the same time. This is the race we don't care about. Once they go to put that binding into our map, that's when we de-duplicate and ensure they agree on who won. This way when you do eventually request an instance we're sure that it only comes from a single, shared binding (which can honor things like `@Singleton`).

Reply:
Similarly, we do the same when linking. Two threads might try to link an unlinked binding at the same time. When they go to replace the unlinked binding with the linked one in our map we make sure they agree on who won.

The gist is that it's rare two threads will be doing this work at the same time so we don't try to de-duplicate it. It would be more expensive to have some fancy locking system than just letting two threads do the same work. We just make sure they agree on who won at the end.